editoast/src/models/rolling_stock_model.rs

```diff
@@ -138,53 +138,78 @@ impl From<model::Error> for Error {
-                err
-            }),
-            None => {
-                let mut err = ValidationErrors::new();
-                err.add(
-                    "effort_curves",
-                    ValidationError::new("effort_curves is required"),
-                );
-                Err(err)
+                    .map(units::second::new)
+                    .is_none()
+                {
+                    let mut error = ValidationError::new("electrical_power_startup_time");
+                    error.message = Some(
+                        "electrical_power_startup_time is required for electric rolling stocks"
+                            .into(),
+                    );
+                    validation_errors.add("effort_curves", error);
+                }
+                if self
+                    .raise_pantograph_time
+                    .flatten()
+                    .map(units::second::new)
+                    .is_none()
+                {
+                    let mut error = ValidationError::new("raise_pantograph_time");
+                    error.message = Some(
+                        "raise_pantograph_time is required for electric rolling stocks".into(),
+                    );
+                    validation_errors.add("effort_curves", error);
+                }
             }
+        } else {
+            validation_errors.add(
+                "effort_curves",
+                ValidationError::new("effort_curves is required"),
+            );
         }
     }
-}
 
-pub fn validate_rolling_stock(
-    effort_curves: &EffortCurves,
-    electrical_power_startup_time: Option<Time>,
-    raise_pantograph_time: Option<Time>,
-) -> std::result::Result<(), ValidationError> {
-    if !effort_curves.is_electric() {
-        return Ok(());
-    }
-    if electrical_power_startup_time.is_none() {
-        let mut error = ValidationError::new("electrical_power_startup_time");
-        error.message =
-            Some("electrical_power_startup_time is required for electric rolling stocks".into());
-        return Err(error);
-    }
-    if raise_pantograph_time.is_none() {
-        let mut error = ValidationError::new("raise_pantograph_time");
-        error.message =
-            Some("raise_pantograph_time is required for electric rolling stocks".into());
-        return Err(error);
+    fn validate_primary_category(&self, validation_errors: &mut ValidationErrors) {
+        if let Some(primary_category) = &self.primary_category {
+            if let Some(other_categories) = &self.other_categories {
+                if other_categories
+                    .iter()
+                    .flatten()
+                    .collect::<Vec<_>>()
+                    .contains(&primary_category)
+                {
+                    let mut error = ValidationError::new("primary_category");
+                    error.message = Some("The primary_category cannot be listed in other_categories for rolling stocks.".into(),);
```

Comment:
```suggestion
                    error.message = Some("The primary_category cannot be listed in other_categories for rolling stocks.".into());
```

I'm actually surprised the formatter didn't catch that. I've seen the formatter having problem with long string but it will [soon](https://doc.rust-lang.org/edition-guide/rust-2024/index.html) (in 2 days 🤞) be over with [the coming 2024's edition](https://doc.rust-lang.org/nightly/edition-guide/rust-2024/rustfmt-formatting-fixes.html#long-strings-dont-prevent-formatting-expressions).

Reply:
Done.